### ISSUE TITLE
Issues with Skript's testing system

### ISSUE THREAD

**Issue by TPGamesNL** [SkriptLang/Skript#3767]
There are currently multiple issues with Skript's testing system, so this thread hopefully summarizes (most of) it.
- There are consistent errors in 1.9.4 and 1.12.2 from the test [`3305-named inventory-item.sk`](https://github.com/SkriptLang/Skript/blob/master/src/test/skript/tests/regressions/3305-named%20inventory-item.sk#L9) ([error](https://github.com/SkriptLang/Skript/runs/1943275005?check_suite_focus=true#step:7:144)).
- Parse errors are hard to find back, since the error telling you that the parsing failed doesn't mention where it failed.
- No tests have to run for the overal result to be a success.
- If all tests are empty (you can try by giving an empty list to [ScriptLoader#L678](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/ScriptLoader.java#L678)), the test is successful.
  - This scenario is unlikely, but I experienced it myself, after I messed something up in `VariableString`, since it's used in every test trigger.
- There are overall not enough tests for the fundamental parts of Skript (e.g. https://github.com/SkriptLang/Skript/pull/3629#issuecomment-748050322 could probably have been prevented with a test).

One thing that I think could be useful for the testing system, is to add a system that generates a syntax tree from a few test scripts, and compares those syntax trees with versions of that syntax tree previously generated for the scripts. I already have a project that can do the syntax tree generation, I'd just need to edit it and make this system work with it.
The benefit of such a system, is that it's easy to add an existing script (without addons), by letting the system generate a syntax tree, and add it to the testing system.

**Comment by LukasmanMHdude** [comment]
also sometimes if you change smth in a function and reload the skript with the function, it will think the function has been deleted. (reloading again fixes the issue)

**Comment by TPGamesNL** [comment]
> also sometimes if you change smth in a function and reload the skript with the function, it will think the function has been deleted. (reloading again fixes the issue)

You should open a new issue for that

**Comment by TPGamesNL** [comment]
> - There are consistent errors in 1.9.4 and 1.12.2 from the test [`3305-named inventory-item.sk`](https://github.com/SkriptLang/Skript/blob/master/src/test/skript/tests/regressions/3305-named%20inventory-item.sk#L9) ([error](https://github.com/SkriptLang/Skript/runs/1943275005?check_suite_focus=true#step:7:144)).
>   - This causes ~75% of the tests to not be ran and thus completely ignored.

When removing the assert that causes these errors, the following tests fail consistently:
```
Failed:
  item comparisons (on 2 environments)
    {_b} is a skeleton skull (it shouldn't be) (on paper-1.12.2)
    {_b} is a skeleton skull (it shouldn't be) (on paper-1.9.4)
  potion effects (on 1 environments)
    the item should have had the speed potion type (on paper-1.9.4)
```

**Comment by TPGamesNL** [comment]
Additionally, some of the tests fail at random ([link](https://github.com/SkriptLang/Skript/pull/3736/commits)):
![image](https://user-images.githubusercontent.com/29547183/109943766-d34f4d00-7cd5-11eb-830b-c2e27c086ca9.png)

[The failed build](https://github.com/SkriptLang/Skript/runs/2024503192#step:7:768):
![image](https://user-images.githubusercontent.com/29547183/109943572-9f742780-7cd5-11eb-9b00-67733a4f08bd.png)

**Comment by TPGamesNL** [comment]
IMO no tests should be ran during Gradle building either

**Comment by TPGamesNL** [comment]
The `AssertionError`s not being caught, is caused by https://github.com/SkriptLang/Skript/blob/d33f96ba9f854a9abf969387c009961ce7f8d913/src/main/java/ch/njol/skript/lang/TriggerItem.java#L97 not catching `AssertionError` (because it's not an `Exception`, it's an `Error`)

**Comment by TheDGOfficial** [comment]
Errors are not supposed to be catched anyway, general practice is to catch
Exception, but if you really want to catch all errors, including weird ones
that probably JVM should handle itself, i.e. OutOfMemoryError, then you
want Throwable, it is the root of all exceptions that can be thrown by
throw statement.

TPGamesNL ***@***.***>, 29 Nis 2021 Per, 15:15 tarihinde şunu
yazdı:

**Comment by TPGamesNL** [comment]
> Errors are not supposed to be catched anyway, general practice is to catch Exception, but if you really want to catch all errors, including weird ones that probably JVM should handle itself, i.e. OutOfMemoryError, then you want Throwable, it is the root of all exceptions that can be thrown by throw statement.

We could catch `Error` (or just `AssertionError` specifically), set the `errored` flag and re-throw it

**Comment by TheLimeGlass** [comment]
I get this error when attempting to use `gradlew skriptTestDev`
```
* What went wrong:
Execution failed for task ':skriptTestDev'.
> Could not set unknown property 'standardInput' for task ':skriptTestDev' of type org.gradle.api.DefaultTask.
```

**Comment by bluelhf** [comment]
> 
> 
> I get this error when attempting to use `gradlew skriptTestDev`
> 
> ```
> * What went wrong:
> Execution failed for task ':skriptTestDev'.
> > Could not set unknown property 'standardInput' for task ':skriptTestDev' of type org.gradle.api.DefaultTask.
> ```

This can be solved by calling createTask with a second argument, `JavaExec`, and omitting the `doFirst` and `javaexec` from the task definition

**Comment by TPGamesNL** [comment]
After #4441, the unit test configs should be updated with each Paper release. Preferably we wouldn't have to do this.

**Comment by TPGamesNL** [comment]
More random fails:
![image](https://user-images.githubusercontent.com/29547183/149927376-94603033-d567-4157-acc6-f6d1c59bfd43.png)
![image](https://user-images.githubusercontent.com/29547183/149927398-2e91eedf-cc1c-4624-ac06-341cc7c266ec.png)

**Comment by TPGamesNL** [comment]
The testing system does not take delays in tests into account, so when using a delay (e.g. wait 1 tick) in a test, there's no guarantee the code after it will run, and therefore it often does not run.

**Comment by TPGamesNL** [comment]
(suggested at https://github.com/SkriptLang/Skript/pull/4845#discussion_r927586147)
We could split the different version tests up into multiple workflows, so they are ran concurrently. 

GH does impose the limit of 20 concurrent jobs (for the whole org), so having one workflow per supported MC version might not be the greatest idea. 

One idea could be the following distribution:
- newest MC version on newest Java version
- oldest MC version on oldest Java version
- the rest of the MC versions, not ran concurrently
